NAAM.md

```diff
@@ -0,0 +1,278 @@
+corresponding to the content hosted by a provider. Once ingested, the indexer nodes facilitate
+lookup of providers over find APIs via the same multihashes.
+
+Naam repurposes the same advertisement structure by specifically crafting entries of advertisement
+such that the indexer node find API can also be repurposed to lookup IPNS records. Naam forms a
+single `EntryChunk` IPLD node, that contains a single multihash calculated as:
+
+* the SHA-256 of IPNS Record Key, i.e. `/ipns/<peer-ID>`.
+
+This approach allows the resolvers to deterministically calculate the multihash used for lookup when
+interacting with the indexer node find APIs.
+
+#### Context ID
+
+The field `ContextID` in a regular advertisement is used as a way to uniquely identify the metadata
+associated to a list of multihashes. Naam uses a fixed value, `/ipni/naam`, as context ID to assure
+that the metadata associated to the published advertisement for the same IPNI key is singular and
+uniquely identifiable.
+
+Using a fixed value as context ID also enables network indexers to quickly differentiate Naam
+advertisements from a typical content advertisement. This opens up future opportunities for
+optimisation when it comes to ingesting Naam advertisements. For example, indexers can pro-actively
+process IPNS records and offer bespoke APIs that understand IPNS record expiry/validity.
+
+#### Metadata
+
+`Metadata` is typically used to convey the list of protocols over which the advertised content is
+retrievable. Further, this field is designed to be extensible: a valid metadata should start with a
+`varint` signifying _some_ protocol ID followed by arbitrary bytes. There are currently three well
+known protocol IDs. The protocol ID itself need not to be known by the indexer nodes. In fact,
+indexers treat metadata as a black box and carry on processing advertisements. Instead, it is meant
+to be a signal to the consumers of `Metadata`, i.e. retrieval clients, as a hint on how to decode
+the remaining bytes.
+
+Naam utilises this extensibility to directly encode IPNS records as metadata. The Naam advertisement
+metadata consists of:
+
+* `ipns` multicodec code `0xe5` as protocol ID,
+* followed by marshalled IPNS Record.
+
+The use of metadata to capture IPNS records enables Naam to also utilise the built-in IPNI mechanism
+for updating metadata to update IPNS records. Because, context ID nor the multihash in Naam
+advertisements ever change for the same IPNS key. For more information,
+see [Updating IPNS Records](#updating-ipns-records).
+
+Note that the IPNS multicodec code `0xe5` is tagged as "namespace" in the CSV table, which could
+signify an IPNS path, not record. There is currently no multicodec code for IPNS records themselves.
+Alternatively, a new multicodec code can be introduced to the
+[multicodec CSV file](https://github.com/multiformats/multicodec/blob/master/table.csv) to avoid
```

Comment:
Off topic: IMO needing to reserve codes in the table is kind of a pain and impediment to development (e.g. see the reluctance around adding a new entry). Would there be any interest in defining some code that just means it's a named thing, so that groups more concerned about iterating/experimenting quickly than saving bytes could just go ahead even if it means losing a few metadata bytes to describe the name of their protocol/system?

Reply:
Potentially; indexers make no assumptions about what's in the metadata. As in, they do not validate it. However, metadata is expected to start with a varint which offers a hint to its consumers about how to read the remaining bytes. 

There are also rules about ordering: metadata entries should be sorted by their protocol ID in ascending order. However, that doesn't apply here because NAAM metadata will only contain a single entry, i.e. `0xe5...`.